### PR TITLE
[NSA] Avoid repeated NSA MQA logits memory queries

### PR DESCRIPTION
**Pull request by YAMY1234**
## Motivation

NSA indexer may call `torch.cuda.mem_get_info` on the large MQA logits path. This is a host-syncing query on a latency-sensitive prefill path. Reported-by: @samuellees

## Modifications

- Cache a per-device MQA logits budget instead of querying free memory on every large prefill.
- Reuse `cu_seqlens_q` in the chunked paged path instead of allocating it inside every chunk loop.

## Accuracy Tests

DeepSeek V3.2 FP4, GB300 TP4, GSM8K full 8-shot (`temperature=0.0`, `max_tokens=14000`, 1311 examples processed):

| Run | Job | Score | Latency | Output throughput |
|---|---:|---:|---:|---:|
| Baseline | 1773283 | 0.961098 | 270.603 s | 885.015 tok/s |
| This PR | 1773284 | 0.962624 | 258.380 s | 916.781 tok/s |

DeepSeek V3.2 FP4, GB300 TP4, GPQA Diamond with `sgl-eval` (`n_repeats=8`, `temperature=1.0`, `top_p=0.95`, `max_tokens=32768`, 198 examples):

| Run | Job | pass@1 | pass@8 | majority@8 | no_answer |
|---|---:|---:|---:|---:|---:|
| Baseline | 1829298 | 84.15% ± 0.44% SEM | 92.93% | 86.36% | 0.44% |
| This PR | 1829300 | 83.14% ± 0.70% SEM | 94.95% | 84.60% | 0.69% |

GPQA is stochastic and the result is close; pass@8 is higher on this PR. GSM8K shows no accuracy regression.

## Speed Tests and Profiling

DeepSeek V3.2 FP4, GB300 TP4, confirmed MQA logits chunk-trigger workload (`isl=65536`, `osl=64`):

| Concurrency | Output throughput delta | Median TTFT delta | P99 TTFT delta | Mean E2EL delta |
|---:|---:|---:|---:|---:|
| 8 | +6.430% | -10.437% | -6.355% | -6.043% |
| 16 | +6.498% | -8.696% | -6.424% | -6.103% |

Torch profiler on 32k/1o/c64 prefill window:

| Run | `nsa_mqa_mem_get_info` calls | Total time |
|---|---:|---:|
| Baseline | 7320 | 677.726 ms |
| This PR | 0 | 0 ms |

Validation also completed for GB300 TP4 8k1k c=64 and 32k1k c=512 sa-bench.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #26114764847](https://github.com/sgl-project/sglang/actions/runs/26114764847)
Latest PR Test (Extra): :warning: **Not enabled** -- add `run-ci-extra` label to opt in.

### CODE REVIEW

**Comment by YAMY1234** [comment]
@Fridge003 @samuellees could you help review this PR? Thanks!

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 1]
```diff
@@ -552,24 +534,45 @@ def _get_topk_paged(
...
+        cached_budget = self._mqa_logits_budget_bytes.get(device_index)
+        if cached_budget is not None:
+            return cached_budget
+
+        total_mem = torch.cuda.get_device_properties(device_index).total_memory
+
+        # Keep the original total-memory guard during CUDA graph capture without
+        # caching it. The first non-capture prefill path will cache the real
+        # free-memory budget below.
+        total_mem_budget = int(total_mem * 0.3)
```
Comment (samuellees) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: Could you please name 0.3 as constants like _MQA_LOGITS_FREE_MEM_FRACTION = 0.5?

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 2]
```diff
@@ -552,24 +534,45 @@ def _get_topk_paged(
...
+        if get_is_capture_mode():
+            return max(1, total_mem_budget)
+
+        # Match the original free-memory guard: logits_bytes * 2 > free_mem.
+        # torch.cuda.mem_get_info synchronizes the host, so cache the result.
+        free_mem, _ = torch.cuda.mem_get_info(device_index)
+        budget_bytes = min(int(free_mem * 0.5), total_mem_budget)
```
Comment (samuellees) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: Same with 0.3

**Review by samuellees — APPROVED** [review]
Nice clean fix. Also left some comments about constant rename^ ^

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py (hunk 1 above)
Comment (YAMY1234) [reply python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: Thanks, Adjusted!

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py (hunk 2 above)
Comment (YAMY1234) [reply python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: Thanks, Adjusted!

**Comment by Fridge003** [comment]
@YAMY1234 Please try benchmarking on GPQA or AIME25 benchmark with sgl-eval: https://github.com/sgl-project/sgl-eval

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 3]
```diff
@@ -552,24 +536,47 @@ def _get_topk_paged(
             topk_result = torch.cat([topk_result, padding], dim=0)
         return topk_result
 
+    def _get_mqa_logits_budget_bytes(self, device_index: int) -> int:
+        cached_budget = self._mqa_logits_budget_bytes.get(device_index)
+        if cached_budget is not None:
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py:563]: I have one question: when we compute the cached budget, what workload is it?
I guess it might be at the first prefill batch during server warmup (before server gets ready). Will this workload be too loose?  The warmup prefill batch should be small, but we wish that the cached budget can be tight enough to avoid smem explosion

**Comment by YAMY1234** [comment]
Ran full GPQA Diamond with `sgl-eval` on DeepSeek V3.2 FP4, GB300 TP4.

Command used: `sgl-eval run gpqa --n-repeats 8 --num-threads 128 --temperature 1.0 --top-p 0.95 --max-tokens 32768`.

Baseline job 1829298:

```text
gpqa rep 1/8: 100%|██████████| 198/198 [48:23<00:00, 14.66s/it, acc=86.36%]
gpqa rep 2/8: 100%|██████████| 198/198 [48:23<00:00, 14.66s/it, acc=83.84%]
gpqa rep 3/8: 100%|██████████| 198/198 [48:23<00:00, 14.66s/it, acc=85.35%]
gpqa rep 4/8: 100%|██████████| 198/198 [48:23<00:00, 14.66s/it, acc=83.84%]
gpqa rep 5/8: 100%|██████████| 198/198 [48:23<00:00, 14.66s/it, acc=82.32%]
gpqa rep 6/8: 100%|██████████| 198/198 [48:23<00:00, 14.66s/it, acc=83.33%]
gpqa rep 7/8: 100%|██████████| 198/198 [48:23<00:00, 14.66s/it, acc=83.84%]
gpqa rep 8/8: 100%|██████████| 198/198 [48:23<00:00, 14.66s/it, acc=84.34%]
gpqa overall: 100%|██████████| 1584/1584 [48:23<00:00,  1.83s/it, acc=84.15%]
== gpqa ==
198 examples x 8 repeats  |  2903.1s  |  3676 tok/s  |  10.7M tokens

* pass@1[avg-of-8]  =  84.15% +/- 1.24% (SEM 0.44%)
  pass@8            =  92.93%
  majority@8        =  86.36%
  no_answer         =  0.44%

Tue May 19 10:07:11 PDT 2026
```

This PR job 1829300:

```text
gpqa rep 1/8: 100%|██████████| 198/198 [48:38<00:00, 14.74s/it, acc=83.33%]
gpqa rep 2/8: 100%|██████████| 198/198 [48:38<00:00, 14.74s/it, acc=83.33%]
gpqa rep 3/8: 100%|██████████| 198/198 [48:38<00:00, 14.74s/it, acc=80.81%]
gpqa rep 4/8: 100%|██████████| 198/198 [48:38<00:00, 14.74s/it, acc=86.36%]
gpqa rep 5/8: 100%|██████████| 198/198 [48:38<00:00, 14.74s/it, acc=84.34%]
gpqa rep 6/8: 100%|██████████| 198/198 [48:38<00:00, 14.74s/it, acc=82.32%]
gpqa rep 7/8: 100%|██████████| 198/198 [48:38<00:00, 14.74s/it, acc=84.34%]
gpqa rep 8/8: 100%|██████████| 198/198 [48:38<00:00, 14.74s/it, acc=80.30%]
gpqa overall: 100%|██████████| 1584/1584 [48:38<00:00,  1.84s/it, acc=83.14%]
== gpqa ==
198 examples x 8 repeats  |  2918.6s  |  3717 tok/s  |  10.8M tokens

* pass@1[avg-of-8]  =  83.14% +/- 1.98% (SEM 0.70%)
  pass@8            =  94.95%
  majority@8        =  84.60%
  no_answer         =  0.69%

Tue May 19 10:06:44 PDT 2026
```

The stochastic GPQA result is mixed but close; pass@8 is higher on this PR, and GSM8K remains unchanged.

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py (hunk 3 above)
Comment (YAMY1234) [reply python/sglang/srt/layers/attention/nsa/nsa_indexer.py:563]: Good point. I updated this in 6f6201f.

The cache is populated only after the static 8M-element skip and only on a non-capture path. Small warmup prefills do not populate it, and CUDA graph capture returns a temporary static guard without caching.

To avoid the first real large prefill making the cached value too loose, the cached free-memory budget is now capped by workload-independent serving headroom derived from `mem_fraction_static` and still by the original total-memory fraction. So even if the first large prefill sees unusually high free memory, the cached budget cannot exceed the serving-time reserve implied by the configured static memory fraction.

**Comment by Fridge003** [comment]
/rerun-test test/registered/dsa_models_e2e/test_dsa_glm5_tp_mtp.py

**Comment by Fridge003** [comment]
/rerun-test test/registered/dsa_models_e2e/test_dsa_glm5_dp_mtp.py

**Comment by Fridge003** [comment]
/rerun-test test/registered/dsa_models_e2e/test_dsa_glm5_tp_mtp.py

**Comment by Fridge003** [comment]
/rerun-test test/registered/dsa_models_e2e/test_dsa_glm5_tp_mtp.py

**Review by Fridge003 — APPROVED** [review]
